docs/docs/dev_docs/contracts/syntax/storage.md

```diff
@@ -1,40 +1,533 @@
+
+This function will throw if the ImmutableSingleton hasn't been initialized.
+
+## `Set<NoteType>`
+
+Set is used for managing a collection of notes. All notes in a set are of the same `NoteType`. But whether these notes all belong to one entity, or are accessible and editable by different entities, is totally up to the developer. Due to our state model, the set is a collection of notes inserted into the data-tree, but notes are never removed from the tree itself, they are only nullified.
+
+#include_code struct  /yarn-project/aztec-nr/aztec/src/state_vars/set.nr rust
+
+And can be added to the `Storage` struct as follows. Here adding a set for a custom note, the TransparentNote (useful for [public -> private communication](./functions.md#public---private)).
+
+#include_code storage_pending_shields /yarn-project/noir-contracts/src/contracts/token_contract/src/main.nr rust
+
+### `new`
+
+The `new` method tells the contract how to operate on the underlying storage.
+
+#include_code new /yarn-project/aztec-nr/aztec/src/state_vars/set.nr rust
+
+Building on before, we can initialize the set as follows:
+
+#include_code storage_pending_shields_init /yarn-project/noir-contracts/src/contracts/token_contract/src/main.nr rust
+
+### `insert`
+
+Allows us to modify the storage by inserting a note into the set.
+
+#include_code insert /yarn-project/aztec-nr/aztec/src/state_vars/set.nr rust
+
+A commitment from the note will be generated, and inserted into data-tree, allowing us to later use in contract interactions. Recall that the content of the note should be shared with the owner to allow them to use it, as mentioned this can be done via an [encrypted log](./events.md#encrypted-events), or offchain via web2, or completely offline.
+
+#include_code insert /yarn-project/aztec-nr/easy-private-state/src/easy_private_state.nr rust
+
+### `insert_from_public`
+
+While we don't support private functions to directly alter public storage, the opposite direction is possible. The `insert_from_public` allow public function to insert notes into private storage. This is very useful when we want to support private function calls that must have been initiated in public, such as shielding or the like.
+
+#include_code insert_from_public /yarn-project/aztec-nr/aztec/src/state_vars/set.nr rust
+
+The usage is rather straight-forward and very similar to using the `insert` method with the difference that this one is called in public functions.
+
+#include_code insert_from_public /yarn-project/noir-contracts/src/contracts/token_contract/src/main.nr rust
+
+### `assert_contains_and_remove`
+
+This function is used to check existence of a note and then remove it without having read the note ahead of time. This can be useful for cases where the user would is providing all the information needed, such as cases where the note was never emitted to the network and thereby available to the wallet.
```

Comment:
```suggestion
This function is used to check existence of a note and then remove it without having read the note ahead of time. This can be useful for cases where the user is providing all the information needed, such as cases where the note was never emitted to the network and thereby available to the wallet.
```